### ISSUE TITLE
Tracking Issue: Frank.Statecharts Spec Pipeline (bidirectional: all formats as input & output)

### ISSUE THREAD

**Issue by panesofglass** [frank-fs/frank#57]
Ultimately, the benefit from something like this should be to define one's own protocol that happens to use HTTP. I think it would be beneficial to show a CE using HTML or [PHTAL](https://tools.ietf.org/html/draft-montoya-phtal-00) that provides state transitions in the results.

---

## Scope Clarification (2026-03-06)

> **This issue covers the full bidirectional spec pipeline** — both the **output direction** (Code → WSD/ALPS/XState/SCXML/smcat) and the **input direction** (WSD/ALPS/XState/SCXML/smcat → F# scaffolding). Users should be able to scaffold an application from any supported design document format, generate or retrieve these documents from a running implementation, and implement without documents then retrieve the generated formats. The **core runtime library** (`Frank.Statecharts` extension, `statefulResource` CE, `IStateMachineStore`, guards, `MailboxProcessor` default) is now tracked in **#87** as a prerequisite issue attached to the v7.3.0 milestone (#80).
>
> **Out of scope**: Orleans virtual actors and other distributed state backends. These are noted as future direction only. The `IStateMachineStore` abstraction enables them without requiring them now.
>
> **Implementation order**: Core library (#87) first (unblocks #76, #77), then this issue (WSD pipeline). Both target v7.3.0.
>
> **Depends on**: #87 (core runtime library must be complete before pipeline work begins).

---

## Sub-Issues (2026-03-07)

This issue has been decomposed into independently trackable sub-issues, ordered by dependency.

**Code generation model**: The "Spec → Code" direction uses **LLM-assisted tooling**, not deterministic template-based generators. Each parser produces a typed representation that becomes structured context for LLM prompts. The LLM generates the F# code (state DUs, transition functions, guard stubs, handler wiring). This is why WSD only needs a parser (#90) — once the spec is parsed into a typed AST, the LLM can generate code from it directly.

### Format Issues (Parser + Generator per format)

Each format issue owns both directions — parsing the format into a typed AST and generating the format from `StateMachineMetadata`. Parser and generator share the same AST types within each issue. WSD is split because the parser (#90) is already underway.

| # | Issue | Scope | Priority | Depends On | Parallel With | Effort |
|---|-------|-------|----------|-----------|---------------|--------|
| #90 | [WSD Parser: F# lexer, parser, and AST](https://github.com/frank-fs/frank/issues/90) | Parser only | **P0** | — | #93, #97, #98, #100 | 2-3 weeks |
| #91 | [WSD Generator + cross-format validator](https://github.com/frank-fs/frank/issues/91) | Generator + validator | **P1** | #90 | #97, #98, #100 | 1-2 weeks |
| #97 | [ALPS Parser + Generator](https://github.com/frank-fs/frank/issues/97) | Bidirectional | **P2** | #87 | #90, #98, #100 | 1-2 weeks |
| #98 | [SCXML Parser + Generator](https://github.com/frank-fs/frank/issues/98) | Bidirectional | **P2** | #87 | #90, #97, #100 | 1-2 weeks |
| #100 | [smcat Parser + Generator](https://github.com/frank-fs/frank/issues/100) | Bidirectional | **P2** | #87 | #90, #97, #98 | 1-2 weeks |
| — | XState JSON | Trivial (both directions) | — | — | — | Handled in #94 |

~~#92 (Spec → Code generator)~~ and ~~#99 (WSD → F# scaffolding)~~ were closed — deterministic code generation is not in scope. The "Spec → Code" direction is LLM-assisted: parsers produce typed representations, LLMs generate the F# code.

### Infrastructure

| # | Issue | Priority | Depends On | Parallel With | Effort |
|---|-------|----------|-----------|---------------|--------|
| #93 | [Automatic ETag generation from state](https://github.com/frank-fs/frank/issues/93) | **P2** | #87 | #90, #91, #97, #98, #100 | 1 week |
| #94 | [frank-cli statechart commands](https://github.com/frank-fs/frank/issues/94) | **P3** | #91, #97, #98, #100 | #93 | 2-3 weeks |

### Dependency Graph

```
#90 WSD Parser ───────────────> #91 WSD Generator + Cross-Validator ┐
(P0, critical path)             (P1)                                 │
                                                                     ├──> #94 frank-cli Commands
#87 Core Runtime ─────────┬───> #97 ALPS Parser + Generator ────────┤    (P3, includes XState JSON
(prerequisite)            ├───> #98 SCXML Parser + Generator ───────┤     both directions)
                          ├───> #100 smcat Parser + Generator ──────┘
                          └───> #93 ETag Generation
                                (P2, parallel — no pipeline dependency)
```

### Format Coverage Matrix

| Format | Input (parse → LLM-assisted codegen) | Output (Code → Spec) |
|--------|--------------------------------------|----------------------|
| **WSD** | #90 (parser) | #91 (generator) |
| **ALPS** | #97 (parser + generator) | #97 |
| **XState JSON** | #94 (trivial) | #94 (trivial) |
| **SCXML** | #98 (parser + generator) | #98 |
| **smcat** | #100 (parser + generator) | #100 |

**Parallel execution opportunities:**
- **Wave 1**: #90 (WSD Parser) ‖ #93 (ETag generation) ‖ #97 (ALPS) ‖ #98 (SCXML) ‖ #100 (smcat) — #90 has no deps; #93/#97/#98/#100 only depend on #87 (prerequisite)
- **Wave 2**: #91 (WSD Generator + cross-validator) — depends on #90
- **Wave 3**: #94 (CLI commands) — depends on all format issues

---

### Problem Statement

Extend Frank's `ResourceBuilder` (or build a companion library) so that HTTP resources expose different representations, allowed methods, and status codes based on **resource state** — implementing HATEOAS through the sequence-diagram-driven API design approach Mike Amundsen advocates.

The [tic-tac-toe app](https://github.com/panesofglass/tic-tac-toe) proves the `MailboxProcessor`-backed agent pattern works. The gap is encoding state machines into ASP.NET Core's request pipeline so the HTTP surface changes per state, including per-user discrimination (e.g., returning `409` when it's not your turn).

---

### Proposed Architecture: WSD → ALPS → XState Pipeline

After evaluating several approaches, a **WSD-as-source-of-truth** pipeline emerges as the strongest option. These three formats model complementary facets of the same system:

| Format | Models | Intentionally Omits |
|--------|--------|---------------------|
| **WSD** | Workflow topology: states, transitions, ordering, parameters | Semantic meaning, HTTP types, data schemas |
| **ALPS** | Vocabulary: semantic descriptors, transition types (`safe`/`unsafe`/`idempotent`), return types | Workflow ordering ([FAQ A.2](http://alps.io/spec/drafts/draft-01.html) explicitly says this) |
| **XState JSON** | Executable state machine: states, transitions, guards, actions, context | HTTP semantics, media types, semantic meaning |
| **SCXML** | W3C standard state machine notation: states, transitions, data model, invoke, history | Semantic meaning, HTTP types, lightweight tooling |
| **smcat** | Lightweight text-based state machine notation with visualization | Guards, data model, semantic meaning |

ALPS FAQ A.2: *"ALPS is not designed to describe workflows or execution paths for a service."* — This is by design. WSD supplies the workflow constraints that ALPS intentionally omits.

```
                 WSD (Source of Truth)
                 States + Transitions + Params + Guards
                        │              │
              ┌─────────┘              └──────────┐
              ▼                                   ▼
       ALPS (Generated)                  XState JSON (Generated)
       • Semantic descriptors            • States & transitions
       • safe/unsafe/idempotent          • Guards & actions  
       • Return types (rt)               • Context from params
       • Parameter schemas               
              │                                   │
              ▼                                   ▼
       app-state-diagram                 Stately Studio
       + API documentation               + Visual validation
       + OpenAPI via eiger                + Simulation / testing
                                          + F# code generation target
```

Both generated artifacts cross-validate each other: every XState event name must exist as an ALPS transition descriptor, every ALPS transition target must exist as an XState state, etc.

#### Mapping Rules: WSD → ALPS

Starting from [Amundsen's onboarding example](https://mamund.site44.com/talks/2018-09-restfest/2018-09-restfest-wsd.pdf):

| WSD Arrow | WSD Meaning | ALPS Type | HTTP |
|-----------|-------------|-----------|------|
| `->` solid forward | Sync call, activates target | `unsafe` | POST |
| `-->` dashed forward | Async/optional, activates target | `unsafe` | POST |
| `->-` solid deactivates | Return from activation | `safe` | GET |
| `-->-` dashed deactivates | Async return | `safe` | GET |

WSD parameters → ALPS semantic descriptors. WSD actor lines → ALPS state descriptors with available transitions nested inside. WSD target states → ALPS `rt` (return type).

#### Guard Extension for WSD

For per-user state discrimination (e.g., tic-tac-toe turns), extend WSD with guard annotations using existing `note` syntax (diagrams still render in websequencediagrams.com):

```
note over WaitingForX: [guard: role=PlayerX]
WaitingForX->+WaitingForO: makeMove(position)
note over WaitingForO: [guard: role=PlayerO]
WaitingForO->+WaitingForX: makeMove(position)
```

Guards flow into ALPS via `ext` elements (ALPS's designated extension point) and into XState as native guards.

---

### Approaches Evaluated

| # | Approach | P(Success) | LLM Cost | Human Time | Maintains Frank Philosophy |
|---|----------|-----------|----------|------------|--------------------------|
| 1 | **Extension library** (`Frank.Statecharts`) | **75%** | $5-75 | 2-4 weeks | ✅ Yes |
| 2 | Bypass ASP.NET Core abstractions | 35% | $50-1500 | 3-6 months | ❌ No |
| 3 | **Spec Kit + prompts** to generate Frank apps | **70%** | $1-10/app | 1-2 weeks setup | ✅ Yes |
| 4 | Generator tool (WSD/XState → Frank) | 55% | $10-100 | 4-8 weeks | ✅ Yes |
| 5a | HTMX/Datastar + SSE for live state projection | 60% | $5-30 | 1-2 weeks | ⚠️ UI-layer only |
| 5b | ~~Orleans virtual actors as state store~~ | ~~65%~~ | ~~$10-50~~ | ~~3-5 weeks~~ | ⚠️ Future direction only |
| 5c | F# Type Provider from WSD spec | 40% | $20-100 | 6-10 weeks | ✅ Very F#-idiomatic |
| 5d | Workflow engine (Elsa/Temporal) | 45% | $5-20 | 2-4 weeks | ❌ Heavy dependency |

> **Note**: Approach 5b (Orleans virtual actors) was evaluated but is **not in scope** for the v7.3.0 milestone. The `IStateMachineStore` abstraction in the core library enables this as a future extension without requiring it now.

#### Why Extension Library (Approach 1) Scores Highest

Frank's design as a thin ASP.NET Core wrapper means all required primitives are available: middleware, `ISession`/`IDistributedCache`, DI, `ClaimsPrincipal`. A `statefulResource` computation expression can wrap the existing `resource` CE. The `MailboxProcessor` from the tic-tac-toe agent translates almost directly — the new layer is the HTTP integration that maps `(user, request, resource state) → (allowed?, status code, representation)`.

#### Performance Impact of MailboxProcessor

- **Per-message overhead**: ~1-5μs (negligible)
- **Per-request with Redis backing**: +~2ms (one `Get` + one `Set`)
- **Memory**: ~few KB per instance; 10K concurrent games ≈ 50-100MB
- **Contention**: `MailboxProcessor` serializes access naturally — ideal for sequential resources like turn-based games; becomes a bottleneck for high-throughput parallel resources

---

### WSD → ALPS → XState/SCXML/smcat Pipeline Feasibility (this issue)

| Phase | Effort | LLM Cost | Probability |
|-------|--------|----------|-------------|
| WSD Parser (F#) — already started in wsd-gen fork | 1-2 weeks | $5-15 | 90% |
| WSD → ALPS Generator | 1 week | $5-10 | 85% |
| WSD → XState JSON Generator | 1 week | $5-10 | 85% |
| WSD → SCXML Generator | 3-5 days | $3-8 | 80% |
| WSD → smcat Generator | 2-3 days | $2-5 | 85% |
| Cross-Validator (ALPS ↔ XState ↔ SCXML) | 2-3 days | $2-5 | 90% |
| smcat Parser | 1 week | $3-8 | 85% |
| Guard syntax extension | 2-3 days | $2-5 | 80% |
| ALPS JSON/XML Parser | 1-2 weeks | $5-10 | 80% |
| SCXML Parser | 1-2 weeks | $5-10 | 80% |
| **Total Pipeline** | **7-12 weeks** | **$35-80** | **72%** |

<details>
<summary>Example: Generated ALPS from Amundsen's onboarding WSD</summary>

```json
{
  "$schema": "https://alps-io.github.io/schemas/alps.json",
  "alps": {
    "version": "1.0",
    "title": "Onboarding API Profile",
    "doc": { "value": "Generated from onboarding.wsd" },
    "descriptor": [
      { "id": "identifier", "type": "semantic" },
      { "id": "name", "type": "semantic" },
      { "id": "email", "type": "semantic" },
      { "id": "region", "type": "semantic" },
      { "id": "discount", "type": "semantic" },

      { "id": "home", "type": "semantic",
        "descriptor": [{ "href": "#startOnboarding" }] },
      { "id": "WIP", "type": "semantic",
        "descriptor": [
          { "href": "#identifier" },
          { "href": "#collectCustomerData" },
          { "href": "#collectAccountData" },
          { "href": "#completeOnboarding" },
          { "href": "#abandonOnboarding" }
        ] },
      { "id": "customerData", "type": "semantic",
        "descriptor": [
          { "href": "#identifier" }, { "href": "#name" },
          { "href": "#email" }, { "href": "#saveToWIP" }
        ] },

      { "id": "startOnboarding", "type": "unsafe", "rt": "#WIP",
        "descriptor": [{ "href": "#identifier" }] },
      { "id": "collectCustomerData", "type": "safe", "rt": "#customerData",
        "descriptor": [
          { "href": "#identifier" }, { "href": "#name" }, { "href": "#email" }
        ] },
      { "id": "saveToWIP", "type": "unsafe", "rt": "#WIP",
        "descriptor": [{ "href": "#identifier" }] },
      { "id": "completeOnboarding", "type": "unsafe", "rt": "#finalizeWIP",
        "descriptor": [{ "href": "#identifier" }] },
      { "id": "goHome", "type": "safe", "rt": "#home" }
    ]
  }
}
```

</details>

<details>
<summary>Example: Generated XState JSON from the same WSD</summary>

```json
{
  "id": "onboarding",
  "initial": "home",
  "states": {
    "home": {
      "on": { "startOnboarding": { "target": "WIP" } }
    },
    "WIP": {
      "on": {
        "collectCustomerData": { "target": "customerData" },
        "collectAccountData": { "target": "accountData" },
        "completeOnboarding": { "target": "finalizeWIP" },
        "abandonOnboarding": { "target": "cancelWIP" }
      }
    },
    "customerData": {
      "on": { "saveToWIP": { "target": "WIP" } }
    },
    "accountData": {
      "on": { "saveToWIP": { "target": "WIP" } }
    },
    "finalizeWIP": {
      "on": { "goHome": { "target": "home" } }
    },
    "cancelWIP": {
      "on": { "goHome": { "target": "home" } }
    }
  }
}
```

</details>

<details>
<summary>Example: Generated F# from XState JSON (LLM-assisted)</summary>

```fsharp
type OnboardingState =
    | Home | WIP | CustomerData
    | AccountData | FinalizeWIP | CancelWIP

type OnboardingEvent =
    | StartOnboarding of identifier: string
    | CollectCustomerData of identifier: string * name: string * email: string
    | SaveToWIP of identifier: string
    | CollectAccountData of identifier: string * region: string * discount: string
    | CompleteOnboarding of identifier: string
    | AbandonOnboarding of identifier: string
    | GoHome

let transition state event =
    match state, event with
    | Home, StartOnboarding _ -> Some WIP
    | WIP, CollectCustomerData _ -> Some CustomerData
    | CustomerData, SaveToWIP _ -> Some WIP
    | WIP, CollectAccountData _ -> Some AccountData
    | AccountData, SaveToWIP _ -> Some WIP
    | WIP, CompleteOnboarding _ -> Some FinalizeWIP
    | FinalizeWIP, GoHome -> Some Home
    | WIP, AbandonOnboarding _ -> Some CancelWIP
    | CancelWIP, GoHome -> Some Home
    | _ -> None
```

</details>

---

### Recommended Implementation Path

> **Prerequisite**: The core `Frank.Statecharts` extension library may be tracked separately (pending feasibility research in kitty-specs/003). It provides the runtime types, `statefulResource` CE, guards, `IStateMachineStore` abstraction, and `MailboxProcessor` default that this pipeline generates code against.

**Phase 1 (Weeks 1-2):** Use Spec Kit to regenerate tic-tac-toe with per-user turn restriction via status codes. Validates the design with minimal investment. *(Out of scope for this issue — separate validation exercise.)*

**Phase 2 (Weeks 3-6):** ~~Extract patterns into `Frank.Statecharts` extension library.~~ → May be tracked in prerequisite issue (pending research).

**Phase 3 (Weeks 7-12):** Build the WSD parser and ALPS/XState/SCXML/smcat generators. WSD parser already partially exists from the [wsd-gen F# fork](https://github.com/panesofglass/wsd-gen/tree/fsharp). **This is the primary deliverable of this issue.**

**~~Phase 4 (Optional):~~** ~~Evaluate Orleans virtual actors as an alternative distributed state store.~~ → Future direction only. Not in v7.3.0 scope.

### Key Design Decisions

1. **State storage abstraction**: Define `IStateMachineStore` early so `MailboxProcessor` can be swapped in the future without changing resource definitions. *(Core library scope.)*
2. **Guard signature**: `ClaimsPrincipal -> 'State -> 'Event -> GuardResult` where `GuardResult` is a DU (`Allowed | NotYourTurn | InvalidTransition | ...`) mapping to specific HTTP status codes. *(Core library scope.)*
3. **Representation strategy**: Amundsen/REST approach favors returning a representation *without* unavailable transitions (filtered affordances) rather than error codes for blocked states
4. **Concurrency model**: `MailboxProcessor` serializes naturally — correct for sequential resources, bottleneck for parallel. Decide if MVP needs parallel states.
5. **Automatic ETag generation from state**: Since `IStateMachineStore` knows the current `('State * 'Context)` for every instance, the middleware can hash the state+context pair and automatically set `ETag` headers on GET responses. This enables conditional request handling (304 Not Modified, 412 Precondition Failed for concurrent mutation) with zero developer effort. The state store already serializes access via `MailboxProcessor`, so optimistic concurrency via `If-Match` is a natural fit — a client that reads state, receives an ETag, and later PUTs with `If-Match` gets 412 if another transition occurred in between. This is a [Webmachine](https://github.com/webmachine/webmachine)-inspired concept (Webmachine handles ETags automatically via `generate_etag` callbacks in its fixed decision tree) adapted to application-level state: the ETag reflects *domain state*, not just response content. See also the [comparison with Webmachine and Freya](https://github.com/frank-fs/frank/blob/master/docs/COMPARISON.md).

### References

- [Mike Amundsen: Using Web Sequence Diagrams with your APIs (RESTFest 2018)](https://mamund.site44.com/talks/2018-09-restfest/2018-09-restfest-wsd.pdf)
- [State Transitions through Sequence Diagrams (F# Advent 2018)](https://wizardsofsmart.wordpress.com/2018/12/05/state-transitions-through-sequence-diagrams/)
- [ALPS Specification](http://alps.io/spec/drafts/draft-01.html)
- [app-state-diagram (ASD)](https://github.com/alps-asd/app-state-diagram) — generates state diagrams from ALPS
- [XState / Stately](https://stately.ai/) — state machine validation and visual editing
- [Restate + XState: Persistent serverless state machines](https://www.restate.dev/blog/persistent-serverless-state-machines-with-xstate-and-restate)
- [State Backed: XState as a backend service](https://docs.statebacked.dev/)
- [SCXML W3C Recommendation](https://www.w3.org/TR/scxml/) — State Chart XML standard
- [smcat (state machine cat)](https://github.com/sverweij/state-machine-cat) — lightweight state machine notation and visualization
- [How is Frank.Statecharts different from Webmachine or Freya?](https://github.com/frank-fs/frank/blob/master/docs/COMPARISON.md) — comparison of application-level vs. protocol-level HTTP state machines
- [RFC 9110 §13: Conditional Requests](https://www.rfc-editor.org/rfc/rfc9110#section-13) — ETag and conditional request semantics

**Comment by panesofglass** [comment]
## Progress Update: #100 (smcat Parser + Generator) complete

**#113** (smcat: migrate parser and generator to shared AST) has been closed via spec 027-smcat-native-annotations.

### What was done:
- Expanded `SmcatMeta` DU with `SmcatStateType of StateKind * SmcatTypeOrigin`, `SmcatTransition of SmcatTransitionKind`, renamed `SmcatActivity` → `SmcatCustomAttribute`
- Generator produces typed AST with `Kind = Initial/Final` on pseudo-states, `SmcatAnnotation` on all nodes
- Serializer consumes `SmcatStateType` for `[type="..."]` attribute emission (Explicit → emit, Inferred → skip) with cross-format fallback
- Parser stores type origin (Explicit/Inferred) and transition semantic roles
- Round-trip fidelity tests with structural equivalence comparison (842 tests pass)

### Updated sub-issue status:
| # | Issue | Status |
|---|-------|--------|
| #111 | Shared Statechart AST | ✅ Done |
| #90 | WSD Parser | ✅ Done |
| #91 | WSD Generator | ✅ Done |
| #97 | ALPS Parser + Generator | ✅ Done |
| #98 | SCXML Parser + Generator | ✅ Done |
| #100 | smcat Parser + Generator | ✅ Done (includes #113) |
| #112 | Cross-format validator | ✅ Done |
| #93 | ETag Generation | ✅ Done |
| #94 | frank-cli Commands | 🔄 In Progress |
| #95 | Production Readiness | ✅ Done |
